lib/Solver/STPBuilder.cpp

```diff
@@ -465,26 +465,29 @@ ExprHandle STPBuilder::getInitialRead(const Array *root, unsigned index) {
   return vc_readExpr(vc, getInitialArray(root), bvConst32(32, index));
 }
 
-::VCExpr STPBuilder::getArrayForUpdate(const Array *root, 
+::VCExpr STPBuilder::getArrayForUpdate(const Array *root,
                                        const UpdateNode *un) {
+  ::VCExpr un_expr;
+  std::vector<const UpdateNode *> update_nodes;
+  for (; un && !_arr_hash.lookupUpdateNodeExpr(un, un_expr);
+       un = un->next.get()) {
+    update_nodes.emplace_back(un);
+  }
   if (!un) {
-      return getInitialArray(root);
+    un_expr = getInitialArray(root);
   }
```

Comment:
```suggestion
  } else {
    un_expr = _arr_hash.lookupUpdateNodeExpr(un, un_expr);
  }
```

We end up with `un_expr` either pointing to the array expr or last relevant cached entry.

Reply:
This has already been done as part of the loop condition, where it is necessary to shortcut the full iteration if it is cached. Am I missing something here?

Reply:
True, I had a different API in mind.